RFC 0001.md

```diff
@@ -0,0 +1,105 @@
+# RFC 0001 — `__torch_function__` for methods of the `torch.Tensor` class.
+## Abstract
+This RFC describes changes necessary to allow `__torch_function__` to be used by methods of `torch.Tensor` in an attempt to make subclassing more accessible to the users of the class. This entails making an API for subclass views public, and a change in the signature of `__torch_function__`.
+
+## Motivation and Scope
+Quoting [[1]], [[2]] and [[3]], the potential goals of this proposal are:
+
+1. Support subclassing `torch.Tensor` in Python
+2. Preserve `Tensor` subclasses when calling `torch` functions on them
+3. Preserve `Tensor` subclasses when calling `numpy` functions on them
+4. Use the NumPy API with PyTorch tensors (i.e. NumPy API calls dispatch to `torch` functions)
+5. Use the PyTorch API with `torch.Tensor`-like objects that are _not_ `Tensor` subclasses
+6. Reuse NumPy ufunc implementations directly from PyTorch
+7. Allow operations on mixed array types, e.g. `tensor + ndarray`
+8. Preserve `Tensor` subclasses when calling `Tensor` methods.
+9. Propagating subclass instances correctly also with operators, using views/slices/indexing/etc.
+10. Preserve subclass attributes when using methods or views/slices/indexing.
+11. A way to insert code that operates on both functions and methods uniformly (so we can write a single function that overrides all operators).
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have a much more streamlined experience. Namely, the following code example will work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors hooks to run whenever methods or operators are called, and to modify the result to their specific use-case, perform logging, or otherwise change the result or the action of the method.
+
+## Detailed Description
+We propose the following signature change to `__torch_funcion__`, to make it match NumPy: [[4]] 
+
+```python
+class SubTensor(torch.Tensor):
+    def __torch_tensor__(self, func, types, args, kwargs):
```

Comment:
You don't say what the old type of `__torch_tensor__` was so I can't tell what the difference is.

@ngoldbaum do you remember why we didn't line up the type exactly with Numpy's type in the beginning?

Reply:
IIRC that was me. Not 100% sure, but I believe I removed (or never added) `types` because it was extra complexity to handle subclasses, and `Tensor` subclasses aren't really a thing today.

Reply:
No reason afaik. That API got hammered out before I started working on the feature so I don’t know why the API is different. It wouldn’t be terribly hard to add `types` to the signature if we wanted to do that.

Reply:
I think it's safe to say we should do this sooner rather than later, even if other parts of this RFC change in their design.

Reply:
As I wrote over in https://github.com/pytorch/pytorch/pull/30730#issuecomment-583262431, my motivation for the `types` argument in NumPy's `__array_function__` was never about subclasses (which I agree are generally awful). My concern was making it as easy and idiomatic as possible for implementers of `__array_function__` to defer to unrecognized types that might also implement an operation. This is generally a good thing for the ecosystem, but people writing special methods tend not to bother, at least for builtin numeric protocols like `__add__` :).

Reply:
> was never about subclasses (which I agree are generally awful)

I kind of disagree. `numpy.ndarray` subclasses give us a bad taste in the mouth because of the history of `numpy.matrix` and other badly written subclasses. But in principle they make sense. Also for PyTorch there's a lot of interest, and the use cases are solid - and letting those users write a whole `torch.Tensor`-like object is definitely not feasible in most cases.